File: solutions/elixir/diamond/1/lib/diamond.ex
# Dedicated to Shree DR.MDD

defmodule Diamond do
  @spec build_shape(char) :: String.t()
  def build_shape(letter) do
    build_pyramid(Enum.to_list(?A..letter), [])
    |> reflect_pyramid()
    |> Enum.map(fn line -> to_string(line) <> "\n" end)
    |> Enum.join("")
  end

  defp build_pyramid([], acc), do: acc

  defp build_pyramid([ch | rest], acc) do
    build_pyramid(rest, expand_pyramid(ch, acc))
  end

  defp build_row(?A), do: 'A'

  defp build_row(ch) do
    [ch] ++ List.duplicate(?\s, 2 * (ch - ?A) - 1) ++ [ch]
  end

  defp expand_row(row), do: [?\s] ++ row ++ [?\s]

  defp expand_pyramid(ch, pyramid) do
    Enum.map(pyramid, &expand_row/1) ++ [build_row(ch)]
  end

  defp reflect_pyramid(pyramid) do
    pyramid ++ Enum.reverse(Enum.take(pyramid, length(pyramid) - 1))
  end
end

File: solutions/erlang/diamond/1/src/diamond.erl
% For my Junko F. Didi and Shree DR.MDD

-module(diamond).
-export([rows/1]).

rows([EndLetter]) ->
    generate_diamond(EndLetter).

generate_diamond(Char) when Char >= $A, Char =< $Z ->
    generate_diamond(Char, 0, 2*(Char-$A)-1, []);
generate_diamond(Char) when Char >= $a, Char =< $z ->
    generate_diamond(Char-$a+$A).

generate_diamond(Char, LR, Mid, Acc) when Mid =< 0 ->
    Line = build_line(Char, LR, Mid),
    Acc1 = [Line|Acc],
    [_|Lower] = lists:reverse(Acc1),
    Acc1 ++ Lower;
generate_diamond(Char, LR, Mid, Acc) ->
    Line = build_line(Char, LR, Mid),
    generate_diamond(Char-1, LR+1, Mid-2, [Line|Acc]).

build_line(Char, LR, Mid) when Mid > 0 ->
    lists:flatten([make_padding(LR), Char, make_padding(Mid), Char, make_padding(LR)]);
build_line(Char, LR, _) ->
    lists:flatten([make_padding(LR), Char, make_padding(LR)]).

make_padding(0) -> "";
make_padding(N) -> [$\s || _ <- lists:seq(1, N)].
